src/newtondynamics_physics/SysNewton.h

```diff
@@ -0,0 +1,196 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2020 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <osp/CommonPhysics.h>
+#include <osp/Active/physics.h>
+#include <osp/Active/ActiveScene.h>
+
+#include <cstdint>
+
+class NewtonBody;
+class NewtonCollision;
+class NewtonWorld;
+
+namespace ospnewton
+{
+
+/**
+ * Component that stores the NewtonWorld, only added to the root node.
+ */
+struct ACompNwtWorld
+{
+    NewtonWorld *m_nwtWorld{nullptr};
+    float m_timeStep{1.0f / 60.0f};
+};
+
+struct ACompNwtBody
+{
+    constexpr ACompNwtBody(NewtonBody const* body) noexcept
+     : m_body(body) { }
+
+    constexpr NewtonBody const* body() const noexcept { return m_body; }
+
+private:
+    NewtonBody const *m_body;
+};
+
+/**
+ * Stores a handle to a NewtonCollision object
+ */
+struct ACompNwtCollider
+{
+    constexpr ACompNwtCollider(NewtonCollision const* collision) noexcept
+     : m_collision(collision) { }
+
+    constexpr NewtonCollision const* collision() const noexcept
+    { return m_collision; }
+
+private:
+    NewtonCollision const *m_collision;
+};
+
+class SysNewton
+{
+
+public:
+
+    static void add_functions(osp::active::ActiveScene &scene);
+    
+    static void update_world(osp::active::ActiveScene& rScene);
+
+    static ACompNwtWorld* try_get_physics_world(osp::active::ActiveScene &rScene);
+
+    /**
+     * Create a Newton TreeCollision from a mesh using those weird triangle
+     * mesh iterators.
+     * @param shape Shape component to store NewtonCollision* into
+     * @param start
+     * @param end
+     */
+    template<class TRIANGLE_IT_T>
+    static void shape_create_tri_mesh_static(
+            osp::active::ActiveScene& rScene, osp::active::ACompShape &rShape,
+            osp::active::ActiveEnt chunkEnt,
+            TRIANGLE_IT_T const& start, TRIANGLE_IT_T const& end);
```

Comment:
TRIANGLE_IT_T && ?

Reply:
I'd probably just rework this whole thing. 'Triangle iterators' here are just a way to turn mesh data into a list of triangles. This is what Newton wants, but probably not what most other physics engines want. A better way to go upon this is just to pass all the mesh data in, and describe it normally with stride, offset, component counts, etc... I feel that something in Magnum::Trade already does this.